### DISCUSSION TITLE
[Proposal] Remove abstract values from classes

### DISCUSSION THREAD

**Discussion by benfurfie** [tailwindlabs/tailwindcss#2327] [Ideas]
One of the best things we've found about using Tailwind – from a commercial and management perspective – has been the ability to rapidly on-board new developers. The opinionated approach means that the class used to define `margin-right: 1.25rem`, for example, is the same on every Tailwind project, provided they've stuck with the convention over configuration philosophy of the project.

Further, because every project following convention uses the same scale, even if a class uses a non-standard size (e.g. w-90), it's extremely quick for the developer to understand that the class applies 360px/22.5rem of width to the element. Similarly, if a developer is applying a new design, they can quickly add a new size to the configuration and every other developer on the project will immediately know what that class does.

That removal of the need to 'learn' what a class means (such as `.button` in semantic CSS, which can render in a million different ways) or source dive, makes it extremely quick to get new developers and contractors up to speed and contributing, and existing developers to rapidly work on code they may not have worked on for months.

However, there is one part of the framework though, that I feel runs counter intuitive to this philosophy – those classes that use abstract values.

This proposal argues that we should be looking to completely remove all but a very limited number of these abstract values from the framework.

## What do I mean by an 'abstract value'?

An abstract value is any that applies a semantic meaning, rather than a mathematical one to the class where there is a scale. Examples of this include:

`.text-2xl`, `.leading-tight`, `.max-w-4xl`, and `.tracking-wide`.

Looking at these classes, it is impossible to tell what value they apply.

They could be replaced, for example, by:

`.text-2xl` => `.text-6` 
`.leading-tight` => `.leading-5` 
`.max-w-4xl` => `.max-w-224` 
`.tracking-wide` => `.tracking-1`†

† Tracking would probably require more thought and discussion due to the different scale used compared to the rest of the framework.

At an absolute minimum, we should be aiming to remove the xl classes, which have no clear scale increment.

## Why would this be a good thing?

As I've mentioned, Tailwind's real value fiscally is the speed at which we can onboard developers who may have never worked on the project before. This would extend it even further by removing the last couple of classes where there is no easy way for developers to quickly understand what is going on without source diving (which is complicated further by the convention of not including the default values in the Tailwind config by default.

Needing to source dive is a code smell, and we should be ideally doing everything we can to minimize the need to do it.

## Is there a precedent?

We've already seen somewhat of a shift toward moving away from abstract classes for things like line-height in Tailwind UI. More often than not, you'll find a spacing-based value rather than the abstract tight, normal, wide style classes.

For example, it's not uncommon to come across classes like `leading-6`, where in the past, we may have used `leading-tight`.

## Any reasons for not implementing it?

The two main reasons I foresee is that it would require a breaking change in order to enable Adam and the team to remove abstract classes entirely, such as those used in `text`, `max-w`, `leading`, and `tracking`, and that it would prevent the use of font sizes that are not divisible by 4.

The only area I would recommend we do not remove these values from is text-sizes between 12px and 18px. Here, the existing `.text-xs`, `.text-sm`, `.text-base`, and `.text-lg` are easy to understand and provide a key level of functionality at a point in the text scale where increments of 2px are critical. However, I would recommend removing `.text-xl`, as it is easily replaced by the `.text-5` class.

As previously mentioned, tracking would also probably need more discussion, as the scales used here are slightly different and not one to one with the rest of the scale.

Previously too, there would have been a risk of bloating the stylesheet; but with PurgeCSS now built in, this is also negated. The only place this wouldn't be the case is in the CDN version. However, this obviously shouldn't be used for production purposes and is already quite bloated anyway.

**Comment by benfurfie** [comment]
I should add, the only change I make to the default configs we use in our projects is the below:

```
fontSize: theme => ({
  xs: '0.75rem',
  sm: '0.875rem',
  base: '1rem',
  lg: '1.125rem',
  xl: '1.25rem',
  ...theme('spacing'),
}),
```

Basically, we remove the $xl classes and inject the spacing values instead, so we can use classes like `.text-10` to output text at 40px.

**Comment by MathGeniusJodie** [comment]
one reason why I think the current approach is fine is that it's more honest, if I set the font size to say 4, the size of the letters isn't 4, it's something else, based on the metrics of the font I use

**Comment by MathGeniusJodie** [comment]
tailwind already has number values for line height https://tailwindcss.com/docs/line-height

**Comment by MathGeniusJodie** [comment]
I would like to add though, "leading" is an inappropriate name for the number values

https://en.wikipedia.org/wiki/Leading

**Comment by MathGeniusJodie** [comment]
when css allows us to set font size and leading like https://seek-oss.github.io/capsize/ , this approach would make sense, until then, handling text size and leading is just gonna have to suck for a while

**Comment by mattmischuk** [comment]
@adamwathan I would like to bump this proposal. If you are doing design right, then you are using the font features to the fullest.

Tailwind as it currently stands is quite punishing if you have a font with a horrid default letter-spacing such as the font [Inter](https://fonts.google.com/specimen/Inter), you see the main issue.

The default letter-spacing is quite awful from a design perspective, so my options are either to do a global css rule or adjust the range values. If this change is added in the next major release of Tailwind that brings the spirit of the law more inline the numeric increment that is used for colors.

I am a weird user as I am a designer who designs in code, and I can tell you objectively the below example is much better from my own experiencing overriding the TW default.

```js
{
  'letterSpacing': {
    '200': '-.075em',
    '300': '-.05em',
    '400': '-.025em',
    '500': '0',
    '600': '.025em',
    '700': '.05em',
  },
}
```
